src/frontends/onnx/frontend/src/op/com.microsoft/matmulnbits.cpp

```diff
@@ -0,0 +1,228 @@
+            const auto transposed_b = std::make_shared<v1::Transpose>(slice_b, transposed_shape);
+
+            // If no zero-points provided - we generate default, depends on data size
+            if (!zero_points.get_node_shared_ptr()) {
+                zero_points = default_zp;
+            }
+            const auto sub_b = std::make_shared<v1::Subtract>(transposed_b, zero_points);
+
+            // Scaling
+            const auto scaled_b = std::make_shared<v1::Multiply>(sub_b, scales);
+
+            // Adding bias if required
+            if (!bias.get_node_shared_ptr()) {
+                b = scaled_b;
+            } else {
+                b = std::make_shared<v1::Add>(scaled_b, bias);
+            }
+        } else {
+            // Transpose matrix
+            const auto transposed_shape =
+                std::make_shared<v0::Constant>(ov::element::i64, Shape{2}, std::vector<int64_t>{1, 0});
+            ov::Output<ov::Node> transposed_b = std::make_shared<v1::Transpose>(converted_b, transposed_shape);
+
+            // If no zero-points provided - we generate default, depends on data size
+            if (!zero_points.get_node_shared_ptr()) {
+                zero_points = default_zp;
+            }
+            const auto sub_b = std::make_shared<v1::Subtract>(transposed_b, zero_points);
+
+            // Scaling
+            const auto scaled_b = std::make_shared<v1::Multiply>(sub_b, scales);
+
+            // Transpose again to make reshaping and slicing
+            transposed_b = std::make_shared<v1::Transpose>(scaled_b, transposed_shape);
+
+            const auto reshaped_b =
+                op::util::reshape(transposed_b,
+                                  ov::Shape{static_cast<size_t>(casted_b_shape[0] / n_blocks_per_col),
+                                            static_cast<size_t>(casted_b_shape[1] * n_blocks_per_col)});
+
+            // Removing unused items in case block is bigger than column count
+            const auto zero_const = std::make_shared<v0::Constant>(ov::element::i32, Shape{1}, 0);
+            const auto one_const = std::make_shared<v0::Constant>(ov::element::i32, Shape{1}, 1);
+            const auto elements_const =
+                std::make_shared<v0::Constant>(ov::element::i32, Shape{1}, static_cast<int32_t>(K));
+            const auto axis_const = std::make_shared<v0::Constant>(ov::element::i32, Shape{1}, 1);
+            const auto slice_b =
+                std::make_shared<v8::Slice>(reshaped_b, zero_const, elements_const, one_const, axis_const);
+
+            // Transpose again
+            transposed_b = std::make_shared<v1::Transpose>(slice_b, transposed_shape);
+
+            // Adding bias if required
+            if (!bias.get_node_shared_ptr()) {
+                b = transposed_b;
+            } else {
+                b = std::make_shared<v1::Add>(transposed_b, bias);
+            }
+        }
+    }
+
+    return {std::make_shared<v0::MatMul>(a, b)};
```

Comment:
can you try to use `transpose_a` and `transpose_b` attributes of `MatMul` to eliminate Transpose nodes in this decompositions?

Reply:
Let me try move this return to another place... I'll get back...

Reply:
Looks like it is possible to remove at least one Transponse operation in case no bias provided for n_blocks_per_col>1.